Suppress log warnings when a sensor group has non numeric members

mill1000 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
None

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Suppress log warnings from sensor groups that have non-numeric members and have been configured to ignore said non-numeric values

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #87924

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

mill1000 committed:
Suppress log warnings when a sensor group has non numeric members and "ignore non numeric" option is selected.

mill1000 commented:
Oops looks like there's some test updates I didn't catch.

mill1000 committed:
Fix and update test cases for sensor group to better test the ignore non numeric function

mill1000 committed:
Fix gibberish in test docstring

teharris1 reviewed (commented):
Is this a breaking change? Given that the `mode` parameter has now been redefined as `ignore_non_numeric` with different logic, is it likely that a `template` is now assuming the `mode` logic?

mill1000 commented:
> Is this a breaking change? Given that the `mode` parameter has now been redefined as `ignore_non_numeric` with different logic, ...

Hm...not sure I follow. With regards to the `mode` parameter the logic is unchanged (except for the log warning).
```
self.mode = all if mode is False else any
```
is equivalent to 
```
validate = any if self._ignore_non_numeric else all
```

> ... is it likely that a `template` is now assuming the `mode` logic?
I don't understand how a `template` (as in template sensor) comes into play here. Can someone manually create a sensor group inside a template?

gjohansson-ST commented on tests/components/group/test_sensor.py:
```diff
@@ -243,15 +243,15 @@ async def test_reload(hass: HomeAssistant) -> None:
             "platform": GROUP_DOMAIN,
             "name": "test_failure",
-            "type": "min",
+            "type": "max",
```
Not that it matters but not sure why you're changing this?

mill1000 replied:
I changed the type to "max" because it makes the asserts at line 273 and line 281 expect different values which I felt slightly improved the test case.

gjohansson-ST reviewed (changes requested):
Some initial comments

mill1000 commented on homeassistant/components/group/sensor.py:
```diff
@@ -324,26 +324,32 @@ def async_update_group_state(self) -> None:
                     sensor_values.append((entity_id, float(state.state), state))
                     if entity_id in self._state_incorrect:
                         self._state_incorrect.remove(entity_id)
+                    valid_states.append(True)
```
This line was relocated from line 338 in order to eliminate the continue on line 337. Which in my opinion makes the loop easier to digest.

Adding `True` to the `valid_states` is used to determine if the group should be marked an available (line 344). If anything, the statement on line 329 `valid_states.append(False)` is unnecessary as `any()` will return `False` on an empty list.

mill1000 committed:
Update tests/components/group/test_sensor.py

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

mill1000 commented:
That CI failure seems to be unrelated to the change

gjohansson-ST commented on homeassistant/components/group/sensor.py:
```diff
@@ -303,7 +303,7 @@ def __init__(
         self._attr_extra_state_attributes = {ATTR_ENTITY_ID: entity_ids}
         self._attr_unique_id = unique_id
-        self.mode = all if mode is False else any
+        self._ignore_non_numeric = ignore_non_numeric
```
How about we keep both and just use the new book for testing if we should log?

mill1000 replied:
Alright

mill1000 committed:
Restore mode members per review

gjohansson-ST commented on homeassistant/components/group/sensor.py:
```diff
@@ -318,7 +318,8 @@ def __init__(
         self._attr_unique_id = unique_id
-        self.mode = all if mode is False else any
+        self._ignore_non_numeric = ignore_non_numeric
+        self._mode = all if ignore_non_numeric is False else any
```
```suggestion
        self.mode = all if ignore_non_numeric is False else any
```

mill1000 replied:
If I may, what's the attachment to `mode`? The only tangentially related references I can find are in the [legacy](https://www.home-assistant.io/integrations/group/#old-style-groups) `Group` class which I believe to be whole independent from `SensorGroup` and it's parent `GroupEntity` classes.

gjohansson-ST replied:
From outside of sensor group nothing I believe more than this is much based from the `min_max` integration. 

I just prefer to leave it as it was as I don't see why this should change.

mill1000 replied:
Ok

gjohansson-ST commented on tests/components/group/test_sensor.py:
```diff
@@ -245,15 +245,15 @@ async def test_reload(hass: HomeAssistant) -> None:
     config = {
         SENSOR_DOMAIN: {
             "platform": GROUP_DOMAIN,
             "name": "test_failure",
```
It's not failing so set a better name

mill1000 replied:
Done

gjohansson-ST commented on tests/components/group/test_sensor.py:
```diff
@@ -266,24 +266,63 @@ async def test_sensor_incorrect_state(
 
     state = hass.states.get("sensor.test_failure")
+    assert state.state == "17.0"
+
```
Use caplog to assert there is no logging

mill1000 replied:
Done

gjohansson-ST commented on tests/components/group/test_sensor.py:
```diff
@@ -266,24 +266,63 @@ async def test_sensor_incorrect_state(
+    assert state.state == "20.0"
+
+
+async def test_sensor_non_numeric(
```
Change the name similar to the test prior this

mill1000 replied:
Done

mill1000 committed:
Address review comments.

- Restore `_mode` to original name
- Update test case and group names
- Assert that non numeric doesn't generate any warning in the logs

gjohansson-ST commented on tests/components/group/test_sensor.py:
```diff
@@ -245,16 +245,58 @@ async def test_reload(hass: HomeAssistant) -> None:
+    state = hass.states.get("sensor.test_ignore_non_numeric")
+    assert state.state == "17.0"
+    assert (
+        "Unable to use state. Only numerical states are supported, entity sensor.test_2 with value string excluded from calculation"
```
```suggestion
        "Unable to use state. Only numerical states are supported,"
```
Let's be less explicit so we ensure to capture

gjohansson-ST commented on tests/components/group/test_sensor.py:
```diff
@@ -245,16 +245,58 @@ async def test_reload(hass: HomeAssistant) -> None:
+    assert state.state == "20.0"
+
+
 async def test_sensor_incorrect_state(
```
```suggestion
async def test_sensor_incorrect_state_with_not_ignore_non_numeric(
```

gjohansson-ST commented on tests/components/group/test_sensor.py:
```diff
@@ -266,24 +308,25 @@ async def test_sensor_incorrect_state(
-    assert state.state == "15.3"
+    assert state.state == "unknown"
     assert (
         "Unable to use state. Only numerical states are supported, entity sensor.test_2 with value string excluded from calculation"
```
```suggestion
        "Unable to use state. Only numerical states are supported"
```
Let's use same as previous test

gjohansson-ST reviewed (changes requested):
Some minor changes in tests then this should be fine to go 👍

mill1000 committed:
Apply suggestions from code review

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

mill1000 committed:
Fix formatting violations

gjohansson-ST reviewed (approved):
Thanks @mill1000 👍

mill1000 commented:
Thank you!
